ksqldb-engine/src/main/java/io/confluent/ksql/utilization/PersistentQuerySaturationMetrics.java

```diff
@@ -0,0 +1,334 @@
+
+    private void reportQuerySaturation(
+        final Instant now,
+        final double saturation,
+        final QueryId queryId,
+        final MetricsReporter reporter
+    ) {
+      LOGGER.info("Reporting query saturation {} for {}", saturation, queryId);
+      reporter.report(ImmutableList.of(
+          new DataPoint(
+              now,
+              QUERY_SATURATION,
+              saturation,
+              ImmutableMap.of(QUERY_ID, queryId.toString())
+          )
+      ));
+    }
+
+    private Map<String, Map<String, Metric>> metricsByThread(final KafkaStreams kafkaStreams) {
+      return kafkaStreams.metrics().values().stream()
+          .filter(m -> m.metricName().group().equals(STREAMS_THREAD_METRICS_GROUP))
+          .collect(Collectors.groupingBy(
+              m -> m.metricName().tags().get(THREAD_ID),
+              Collectors.toMap(m -> m.metricName().name(), m -> m, (a, b) -> a)
+          ));
+    }
+
+    private Optional<Double> measure(
+        final Instant now,
+        final KafkaStreams kafkaStreams,
+        final MetricsReporter reporter
+    ) {
+      final Map<String, Map<String, Metric>> byThread = metricsByThread(kafkaStreams);
+      Optional<Double> saturation = Optional.of(0.0);
+      for (final Map.Entry<String, Map<String, Metric>> entry : byThread.entrySet()) {
+        final String threadName = entry.getKey();
+        final Map<String, Metric> metricsForThread = entry.getValue();
+        if (!metricsForThread.containsKey(STREAMS_TOTAL_BLOCKED_TIME)
+            || !metricsForThread.containsKey(STREAMS_THREAD_START_TIME)) {
+          LOGGER.info("Missing required metrics for thread: {}", threadName);
+          continue;
+        }
+        final double totalBlocked =
+            (double) metricsForThread.get(STREAMS_TOTAL_BLOCKED_TIME).metricValue();
+        final long startTime =
+            (long) metricsForThread.get(STREAMS_THREAD_START_TIME).metricValue();
+        final ThreadSaturation threadSaturation = perThreadSaturation.computeIfAbsent(
+            threadName,
+            k -> {
+              LOGGER.debug("Adding saturation for new thread: {}", k);
+              return new ThreadSaturation(threadName, startTime, window, sampleMargin);
+            }
+        );
+        LOGGER.debug("Record thread {} sample {} {}", threadName, totalBlocked, startTime);
+        final BlockedTimeSample blockedTimeSample = new BlockedTimeSample(now, totalBlocked);
+        final Optional<Double> measured = threadSaturation.measure(now, blockedTimeSample);
+        LOGGER.debug(
+            "Measured value for thread {}: {}",
+            threadName,
+            (measured.map(Object::toString).orElse(""))
+        );
+        measured.ifPresent(m -> reportThreadSaturation(now, m, threadName, reporter));
+        saturation = compareSaturation(saturation, measured) > 0 ? saturation : measured;
+      }
+      saturation.ifPresent(s -> reportQuerySaturation(now, s, queryId, reporter));
```

Comment:
It may be a duplicate comment with @wcarlson5 but with the shared runtime the mapping from queries -> threads is no longer one to many, but many to many, so we may end up double maxing the thread saturation for some queries? Since now it is just taking the max that maybe fine, but if we ever change it to a different equation we may end up duplicating the measurements.

Reply:
Also with the new shared runtime, using `max` over all the threads that host at least one task of the query may be largely exaggerating the saturation metrics: e.g. let's say a single thread hosting some task from query A and some task from query B, and when query B's workload becomes a beast to drag down that thread, query A's saturation metric would also report as "full". This would be bad.. 

In the new shared runtime, there's no good way to tell how much of the saturation of a thread was contributed by one of its hosted query than the other, so a second least approach may be to just consider averaging over the threads.. thoughts @lct45 @wcarlson5

Reply:
> let's say a single thread hosting some task from query A and some task from query B, and when query B's workload becomes a beast to drag down that thread, query A's saturation metric would also report as "full". This would be bad..

I agree that's what would happen, but I think that's actually what we want. The saturation metric should tell you whether a given query is bottlenecked or not - not what % of the capacity that one query is using. The latter is a more challenging problem (I have an idea here around measuring the average record processing time and multiplying by the observed throughput into the query's source partitions) that we're not aiming to solve with this patch. I can see why this might be confusing though - we could leave the query-level metric internal for now.

Reply:
Cool, let's table that for now and merge as-is. We can see how it works well or not in production for monitoring purposes and decide whether we want to do something differently. I was concerned that we may exaggerating the saturation situation and hence put too much monitoring noise to ourselves, where a query that was only partially saturated on one of the hosting threads but not on others would be shown as 100% saturated. But that could be totally over-thinking too :)